rpcserver.go

```diff
@@ -2728,11 +2727,68 @@ func (r *rpcServer) GetNetworkInfo(ctx context.Context,
 		minChannelSize = 0
 	}
 
-	// TODO(roasbeef): graph diameter
+	// Calculate graph diameter: Diameter = maximum length of shortest path
+	// in graph between two nodes.
+	var diam int
+
+	// We'll run through all the nodes in graph, find farthest node from
+	// each, and take the maximum of that.
+	if err := graph.ForEachNode(nil, func(tx *bolt.Tx, SourceNode *channeldb.LightningNode) error {
+
+		// This will store the farthest node
+		var maxdist = -1
+		current := []*channeldb.LightningNode{SourceNode}
+		visited := map[string]bool{}
+
+		visited[hex.EncodeToString(SourceNode.PubKey.SerializeCompressed())] = true
+
+		// current at all times will store the current layer of nodes we
+		// have encountered in BFS traversal. next will denote the (partially)
+		// formed next layer.
+
+		for len(current) > 0 {
+			// Each layer adds 1 to the farthest node distance from SourceNode
+			maxdist++
+			next := []*channeldb.LightningNode{}
```

Comment:
IMO, the code would be clearer if we just used an _explicit_ queue rather than re-assignign this slice variable. We can re-use a `[]*channeldb.LightnignNode` as the queue. 

To append to the queue we'd do: 
```
nextNodes = append(nextNodes, newNode)
```

To pop from the queue, we'd do: 
```
nextNode := nextNodes[0]
nextNodes[0] = nil
nextNodes = nextNodes[1:]
```

Reply:
Yeah, but I don’t really need a queue for this, since that would require maintaining a seperate distance array etc., layer by layer traversal allows to circumvent that(while maintaining readability IMO).

Reply:
Popping doesn't copy the structure. It is just a re-slice, which moves a pointer location.

Reply:
Ah, I see. Not very experienced with Go. Will fix everything soon. :)